src/ring_theory/polynomial_algebra.lean

```diff
@@ -0,0 +1,342 @@
+    refl },
+  map_add' := by {
+    intros, unfold to_fun_linear_right,
+    congr, simp only [linear_map.coe_mk],
+    unfold to_fun finsupp.sum monomial,
+    simp_rw [← finset.sum_add_distrib, ← finsupp.single_add, ← add_mul],
+    refl } }
+
+/--
+(Implementation detail).
+The function underlying `A ⊗[R] polynomial R →ₐ[R] polynomial A`,
+as a linear map.
+-/
+def to_fun_linear : A ⊗[R] polynomial R →ₗ[R] polynomial A :=
+tensor_product.lift (to_fun_bilinear R A)
+
+-- We apparently need to provide the decidable instance here
+-- in order to successfully rewrite by this lemma.
+lemma to_fun_linear_mul_tmul_mul_aux_1
+  (p : polynomial R) (k : ℕ) (h : decidable (¬p.coeff k = 0)) (a : A) :
+  ite (¬coeff p k = 0) (a * (algebra_map R A) (coeff p k)) 0 = a * (algebra_map R A) (coeff p k) :=
+begin
+  haveI := h,
+  by_cases w : ¬p.coeff k = 0,
+  { simp [w], },
+  { simp [classical.not_not.1 w], }
+end
+
+lemma to_fun_linear_mul_tmul_mul_aux_2 (k : ℕ) (a₁ a₂ : A) (p₁ p₂ : polynomial R) :
+  a₁ * a₂ * (algebra_map R A) ((p₁ * p₂).coeff k) =
+    (finset.nat.antidiagonal k).sum
+      (λ x, a₁ * (algebra_map R A) (coeff p₁ x.1) * (a₂ * (algebra_map R A) (coeff p₂ x.2))) :=
+begin
+  simp_rw [mul_assoc, algebra.commutes, ←finset.mul_sum, mul_assoc, ←finset.mul_sum],
+  congr,
+  simp_rw [algebra.commutes (coeff p₂ _), coeff_mul, ring_hom.map_sum, ring_hom.map_mul],
+end
+
+lemma to_fun_linear_mul_tmul_mul (a₁ a₂ : A) (p₁ p₂ : polynomial R) :
+  (to_fun_linear R A) ((a₁ * a₂) ⊗ₜ[R] p₁ * p₂) =
+    (to_fun_linear R A) (a₁ ⊗ₜ[R] p₁) * (to_fun_linear R A) (a₂ ⊗ₜ[R] p₂) :=
+begin
+  dsimp [to_fun_linear],
+  simp only [lift.tmul],
+  dsimp [to_fun_bilinear, to_fun_linear_right, to_fun],
+  ext k,
+  simp_rw [coeff_sum, coeff_single, finsupp.sum,
+    finset.sum_ite_eq', finsupp.mem_support_iff, ne.def, coeff_mul, finset_sum_coeff, coeff_single,
+    finset.sum_ite_eq', finsupp.mem_support_iff, ne.def,
+    mul_ite, mul_zero, ite_mul, zero_mul, apply_eq_coeff],
+  simp_rw [ite_mul_zero_left (¬coeff p₁ _ = 0) (a₁ * (algebra_map R A) (coeff p₁ _))],
+  simp_rw [ite_mul_zero_right (¬coeff p₂ _ = 0) _ (_ * _)],
+  simp_rw [to_fun_linear_mul_tmul_mul_aux_1, to_fun_linear_mul_tmul_mul_aux_2],
+end
+
+lemma to_fun_linear_algebra_map_tmul_one (r : R) :
+  (to_fun_linear R A) ((algebra_map R A) r ⊗ₜ[R] 1) = (algebra_map R (polynomial A)) r :=
+begin
+  dsimp [to_fun_linear],
+  simp only [lift.tmul],
+  dsimp [to_fun_bilinear, to_fun_linear_right, to_fun],
+  rw [← C_1, ←monomial_zero_left, finsupp.sum_single_index],
+  { simp, refl, },
+  { simp, },
+end
+
+/--
+(Implementation detail).
+The algebra homomorphism `A ⊗[R] polynomial R →ₐ[R] polynomial A`.
+-/
+-- TODO This takes about 8 seconds to elaborate. Suggestions welcome.
+def to_fun_alg_hom : A ⊗[R] polynomial R →ₐ[R] polynomial A :=
+alg_hom_of_linear_map_tensor_product
+  (to_fun_linear R A)
+  (to_fun_linear_mul_tmul_mul R A)
+  (to_fun_linear_algebra_map_tmul_one R A)
```

Comment:
Note the very slow elaboration here. Suggestions welcome.

Reply:
Did you investigate which part takes time? For instance, what happens if you take each of the last three lines in turn and replace it by the fully elaborated version?

Reply:
Maybe I'm not thinking straight at the moment, but I'm actually not sure what you're suggesting I try here. For each of those three lines, they are already "maximally abstracted" as far as I can see. I guess I could replace each in turn by `sorry` and see how that affects compilation time.

Reply:
Replacing `(to_fun_linear R A)` with `sorry` just causes it to break (quickly). Replacing either the 2nd last or last line with sorry shaves off about 2.5s from the elaboration time.

Reply:
I wasn't suggesting to replace them with sorry, but with whatever huge term you see when looking at the result with `set_option pp.all true`. That's what elaboration is about: turning human readable input into something having details spelled out, including implicit and classes instances.

Reply:
Of course the idea is not to keep these horrible terms, only to answer the question: which part is long to elaborate.

Reply:
Copying and pasting the output from `pp.all` (and moving the arguments back before the colon, which makes very little difference), reduces elaboration time to 106ms.

* Giving it the implicit arguments of `alg_hom_of_linear_map_tensor_product` but still elaborating the three explicit arguments takes 1.56s.

* Instead making it work out the implicit arguments of `alg_hom_of_linear_map_tensor_product` increases this to 4.71s.

* Making it do both (that is, providing a fully elaborated return type, but the usual definition) brings it up to 5.65s. (So not so far from being cumulative)

* Giving it a fully elaborated definition, but making it elaborate the return type, takes 1.65s. 

* Making it do everything takes 7.4s.

So the bulk of the elaboration time is in finding the implicit arguments in the definition, but there's a fair chunk also in the return type, and in the explicit arguments.

Reply:
Ooof... so now the question is: what is the best compromise.

Reply:
I experimented with making all the type arguments of `alg_hom_of_linear_map_tensor_product` explicit. This reduces elaboration time here from 7.4s to 6.7s --- so sadly not much.

The next option is unfortunately to provide some of the typeclass instances by hand...

Reply:
Is there a way to measure how long typeclass inference itself is taking?

When I turn on `set_option trace.class_instances true`, I get 65 lines reading "[class_instances]  class-instance resolution trace", and the next lines are:

```
[class_instances] (0) ?x_0 : comm_ring.{u_1} R := _inst_1
[class_instances] (0) ?x_0 : semiring.{u_1} R := @polynomial.semiring.{?u_0} ?x_1 ?x_2
[class_instances] (0) ?x_0 : add_comm_group.{u_2} A := @module.restrict_scalars.add_comm_group.{?u_0 ?u_1 ?u_2} ?x_1 ?x_2 ?x_3 ?x_4
[class_instances] (0) ?x_0 : add_comm_group.{u_1} (@polynomial.{u_1} R (@ring.to_semiring.{u_1} R (@comm_ring.to_ring.{u_1} R _inst_1))) := @module.restrict_scalars.add_comm_group.{?u_0 ?u_1 ?u_2} ?x_1 ?x_2 ?x_3 ?x_4
[class_instances] (0) ?x_5 : ring.{u_1} R := _inst_2
[class_instances] (0) ?x_0 : @module.{u_1 u_2} R A (@comm_ring.to_ring.{u_1} R _inst_1) (@ring.to_add_comm_group.{u_2} A _inst_2) := @polynomial.semimodule.{?u_0} ?x_1 ?x_2
[class_instances] (0) ?x_98 : comm_semiring.{u_1} R := @polynomial.comm_semiring.{?u_41} ?x_99 ?x_100
[class_instances] (0) ?x_101 : semiring.{u_2} A := @polynomial.semiring.{?u_42} ?x_102 ?x_103
[class_instances] (0) ?x_0 : @module.{u_1 u_1} R (@polynomial.{u_1} R (@ring.to_semiring.{u_1} R (@comm_ring.to_ring.{u_1} R _inst_1)))
[class_instances] (0) ?x_0 : semiring.{u_2} A := @polynomial.semiring.{?u_0} ?x_1 ?x_2
[class_instances] (0) ?x_0 : comm_semiring.{u_1} R := @polynomial.comm_semiring.{?u_0} ?x_1 ?x_2
[class_instances] (0) ?x_0 : semiring.{(max u_2 u_1)}
[class_instances] (0) ?x_19 : ring.{u_2} A := _inst_2
[class_instances] (0) ?x_20 : ring.{u_1} (@polynomial.{u_1} R (@ring.to_semiring.{u_1} R (@comm_ring.to_ring.{u_1} R _inst_1))) := _inst_2
[class_instances] (0) ?x_23 : ring.{u_1} R := _inst_2
[class_instances] (0) ?x_26 : @algebra.{u_1 u_2} R A (@comm_ring.to_comm_semiring.{u_1} R _inst_1) (@ring.to_semiring.{u_2} A _inst_2) := _inst_3
[class_instances] (0) ?x_27 : @algebra.{u_1 u_1} R (@polynomial.{u_1} R (@ring.to_semiring.{u_1} R (@comm_ring.to_ring.{u_1} R _inst_1)))
[class_instances] (0) ?x_33 : ring.{u_1} R := _inst_2
[class_instances] (0) ?x_36 : comm_ring.{u_1} R := _inst_1
[class_instances] (0) ?x_0 : semiring.{u_2} (@polynomial.{u_2} A (@ring.to_semiring.{u_2} A _inst_2)) := @polynomial.semiring.{?u_0} ?x_1 ?x_2
[class_instances] (0) ?x_0 : @algebra.{u_1 (max u_2 u_1)} R
[class_instances] (0) ?x_43 : ring.{u_2} A := _inst_2
[class_instances] (0) ?x_44 : ring.{u_1} (@polynomial.{u_1} R (@ring.to_semiring.{u_1} R (@comm_ring.to_ring.{u_1} R _inst_1))) := _inst_2
[class_instances] (0) ?x_47 : ring.{u_1} R := _inst_2
[class_instances] (0) ?x_50 : @algebra.{u_1 u_2} R A (@comm_ring.to_comm_semiring.{u_1} R _inst_1) (@ring.to_semiring.{u_2} A _inst_2) := _inst_3
[class_instances] (0) ?x_51 : @algebra.{u_1 u_1} R (@polynomial.{u_1} R (@ring.to_semiring.{u_1} R (@comm_ring.to_ring.{u_1} R _inst_1)))
[class_instances] (0) ?x_57 : ring.{u_1} R := _inst_2
[class_instances] (0) ?x_60 : comm_ring.{u_1} R := _inst_1
[class_instances] (0) ?x_0 : @algebra.{u_1 u_2} R (@polynomial.{u_2} A (@ring.to_semiring.{u_2} A _inst_2))
[class_instances] (0) ?x_6 : ring.{u_2} A := _inst_2
[class_instances] (0) ?x_7 : comm_ring.{u_1} R := _inst_1
[class_instances] (0) ?x_0 : ring.{u_2} A := _inst_2
[class_instances] (0) ?x_0 : ring.{u_1} (@polynomial.{u_1} R (@ring.to_semiring.{u_1} R (@comm_ring.to_ring.{u_1} R _inst_1))) := _inst_2
[class_instances] (0) ?x_3 : ring.{u_1} R := _inst_2
[class_instances] (0) ?x_0 : @algebra.{u_1 u_2} R A (@comm_ring.to_comm_semiring.{u_1} R _inst_1) (@ring.to_semiring.{u_2} A _inst_2) := _inst_3
[class_instances] (0) ?x_0 : @algebra.{u_1 u_1} R (@polynomial.{u_1} R (@ring.to_semiring.{u_1} R (@comm_ring.to_ring.{u_1} R _inst_1)))
[class_instances] (0) ?x_6 : ring.{u_1} R := _inst_2
[class_instances] (0) ?x_9 : comm_ring.{u_1} R := _inst_1
[class_instances] (0) ?x_0 : ring.{u_2} (@polynomial.{u_2} A (@ring.to_semiring.{u_2} A _inst_2)) := _inst_2
[class_instances] (0) ?x_3 : ring.{u_2} A := _inst_2
[class_instances] (0) ?x_0 : comm_ring.{u_1} R := _inst_1
[class_instances] (0) ?x_0 : ring.{u_2} A := _inst_2
[class_instances] (0) ?x_0 : @algebra.{u_1 u_2} R A (@comm_ring.to_comm_semiring.{u_1} R _inst_1) (@ring.to_semiring.{u_2} A _inst_2) := _inst_3
[class_instances] (0) ?x_0 : ring.{u_1} (@polynomial.{u_1} R (@ring.to_semiring.{u_1} R (@comm_ring.to_ring.{u_1} R _inst_1))) := _inst_2
[class_instances] (0) ?x_3 : ring.{u_1} R := _inst_2
[class_instances] (0) ?x_0 : @algebra.{u_1 u_1} R (@polynomial.{u_1} R (@ring.to_semiring.{u_1} R (@comm_ring.to_ring.{u_1} R _inst_1)))
[class_instances] (0) ?x_6 : ring.{u_1} R := _inst_2
[class_instances] (0) ?x_9 : comm_ring.{u_1} R := _inst_1
[class_instances] (0) ?x_0 : ring.{u_2} (@polynomial.{u_2} A (@ring.to_semiring.{u_2} A _inst_2)) := _inst_2
[class_instances] (0) ?x_3 : ring.{u_2} A := _inst_2
[class_instances] (0) ?x_0 : @algebra.{u_1 u_2} R (@polynomial.{u_2} A (@ring.to_semiring.{u_2} A _inst_2))
[class_instances] (0) ?x_6 : ring.{u_2} A := _inst_2
[class_instances] (0) ?x_7 : comm_ring.{u_1} R := _inst_1
[class_instances] (0) ?x_0 : ring.{u_2} A := _inst_2
[class_instances] (0) ?x_0 : @algebra.{u_1 u_2} R A (@comm_ring.to_comm_semiring.{u_1} R _inst_1) (@ring.to_semiring.{u_2} A _inst_2) := _inst_3
[class_instances] (0) ?x_0 : comm_ring.{u_1} R := _inst_1
[class_instances] (0) ?x_0 : ring.{u_2} A := _inst_2
[class_instances] (0) ?x_0 : @algebra.{u_1 u_2} R A (@comm_ring.to_comm_semiring.{u_1} R _inst_1) (@ring.to_semiring.{u_2} A _inst_2) := _inst_3
[class_instances] (0) ?x_0 : comm_ring.{u_1} R := _inst_1
[class_instances] (0) ?x_0 : comm_ring.{u_1} R := _inst_1
[class_instances] (0) ?x_0 : ring.{u_2} A := _inst_2
[class_instances] (0) ?x_0 : @algebra.{u_1 u_2} R A (@comm_ring.to_comm_semiring.{u_1} R _inst_1) (@ring.to_semiring.{u_2} A _inst_2) := _inst_3
[class_instances] (0) ?x_0 : comm_ring.{u_1} R := _inst_1
[class_instances] (0) ?x_0 : ring.{u_2} A := _inst_2
[class_instances] (0) ?x_0 : @algebra.{u_1 u_2} R A (@comm_ring.to_comm_semiring.{u_1} R _inst_1) (@ring.to_semiring.{u_2} A _inst_2) := _inst_3
```

Reply:
What about tweaking instances priorities, maybe locally?

Reply:
My attempts to use shortcut instances have failed. No subset of
```
def F : algebra R (polynomial R) := by apply_instance
local attribute [instance, priority 2000] F 
def G : algebra R (polynomial A) := by apply_instance
local attribute [instance, priority 2000] G 
def H : algebra R (A ⊗[R] polynomial R) := by apply_instance
local attribute [instance, priority 2000] H
```
significantly improves elaboration time.

Reply:
The longest typeclass searches (counting lines in `trace class_instances`) are actually for `ring R`, which eventually find `comm_ring.to_ring`. Adding
```
local attribute [priority 2000] comm_ring.to_ring
```
however makes no improvement.

Reply:
Similarly for `local attribute [priority 2000] algebra.tensor_product.algebra`.

Reply:
I propose we declare victory based on Gabriel's patch to Lean, which saves about 1.5s seconds here.